### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-type-parameters] NoInfer triggers false positive

### ISSUE THREAD

**Issue by guillaumebrunerie** [typescript-eslint/typescript-eslint#9766]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=MYewdgzgLgBAZjAvDAPAFQDQD4AUAoGGCAUygDUBDAGwFdiAuGHAN0bQEoksZmQBLACYYCMAOalKtBk06JuAORABJMHGIAndFmGzuAbxgBfANx4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const f = <T,>(
  setValue: (v: T) => void,
  getValue: () => NoInfer<T>,
) => { };
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The type parameter `T` is used twice, in a way that cannot be done without type parameters, so it shouldn’t report any error.

### Actual Result

The error "Type parameter T is used only once.", which is incorrect.
Or is there a different way I am supposed to write such a function?

### Additional Info

Removing `NoInfer` makes the error message go away. I guess this is covered by the "Limitations" section of the doc, but it seems to make this rule pretty fragile.

**Comment by bradzacher** [comment]
This is probably because `NoInfer` is defined as `type NoInfer<T> = intrinsic;` so it doesn't look like it uses the parameter when the rule does its queries.

**Comment by JoshuaKGoldberg** [comment]
Yeah, any reference to `intrinsic` is an immediate _"we have no idea, don't report"_.

**Comment by Josh-Cena** [comment]
> Yeah, any reference to `intrinsic` is an immediate _"we have no idea, don't report"_.

I don't know what you meant exactly by that, but I would expect `type G<T> = intrinsic` to be equivalent to `type G<T> = T`, at least for the existing intrinsic types.

**Comment by auvred** [comment]
This bug is no longer reproducible on `main`. Turns out https://github.com/typescript-eslint/typescript-eslint/pull/9900 fixed it :+1:
